File: src/org/elixir_lang/module/ElixirModuleBuilder.java
package org.elixir_lang.module;

import com.intellij.compiler.CompilerWorkspaceConfiguration;
import com.intellij.ide.util.projectWizard.*;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleType;
import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.projectRoots.SdkTypeId;
import com.intellij.openapi.roots.ModifiableRootModel;
import com.intellij.openapi.util.Condition;
import com.intellij.openapi.util.Pair;
import org.elixir_lang.icons.ElixirIcons;
import org.elixir_lang.sdk.ElixirSdkType;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.util.Collections;

/**
 * Created by zyuyou on 2015/5/26.
 *
 */
public class ElixirModuleBuilder extends JavaModuleBuilder implements ModuleBuilderListener {
  @Override
  public void setupRootModel(ModifiableRootModel rootModel) throws ConfigurationException {
    addListener(this);
    super.setupRootModel(rootModel);
  }

  @Override
  public ModuleType getModuleType() {
    return ElixirModuleType.getInstance();
  }

  @Override
  public boolean isSuitableSdkType(SdkTypeId sdkType) {
    return sdkType == ElixirSdkType.getInstance();
  }

  @Override
  public Icon getNodeIcon() {
    return ElixirIcons.FILE;
  }

  @Override
  public void moduleCreated(@NotNull Module module) {
    CompilerWorkspaceConfiguration.getInstance(module.getProject()).CLEAR_OUTPUT_DIRECTORY = false;
  }

  @Nullable
  @Override
  public ModuleWizardStep modifySettingsStep(@NotNull SettingsStep settingsStep) {
    return new SdkSettingsStep(settingsStep, this, new Condition<SdkTypeId>() {
      @Override
      public boolean value(SdkTypeId sdkTypeId) {
        return isSuitableSdkType(sdkTypeId);
      }
    }){
      @Override
      public void updateDataModel(){
        super.updateDataModel();
        final String path = getContentEntryPath();
        if(path != null){
          setSourcePaths(Collections.singletonList(Pair.create(path + "/lib", "")));
        }
      }
    };
  }
}

File: src/org/elixir_lang/module/ElixirModuleBuilder.kt
package org.elixir_lang.module

import com.intellij.compiler.CompilerWorkspaceConfiguration
import com.intellij.ide.util.projectWizard.*
import com.intellij.openapi.module.Module
import com.intellij.openapi.module.ModuleType
import com.intellij.openapi.options.ConfigurationException
import com.intellij.openapi.projectRoots.SdkTypeId
import com.intellij.openapi.roots.ModifiableRootModel
import com.intellij.openapi.util.Condition
import org.elixir_lang.Icons
import org.elixir_lang.mix.Project
import org.elixir_lang.sdk.elixir.Type.Companion.instance
import javax.swing.Icon

/**
 * Created by zyuyou on 2015/5/26.
 *
 */
class ElixirModuleBuilder : JavaModuleBuilder(), ModuleBuilderListener {
    @kotlin.jvm.Throws(ConfigurationException::class)
    override fun setupRootModel(rootModel: ModifiableRootModel) {
        addListener(this)
        super.setupRootModel(rootModel)
        val contentEntry = rootModel.contentEntries.single()
        val projectDirectory = contentEntry.file!!

        rootModel.removeContentEntry(contentEntry)
        Project.addFolders(rootModel, projectDirectory)
    }

    override fun getModuleType(): ModuleType<*> = ElixirModuleType.getInstance()
    override fun isSuitableSdkType(sdkType: SdkTypeId): Boolean = sdkType === instance
    override fun getNodeIcon(): Icon = Icons.LANGUAGE

    override fun moduleCreated(module: Module) {
        CompilerWorkspaceConfiguration.getInstance(module.project).CLEAR_OUTPUT_DIRECTORY = false
    }

    override fun modifySettingsStep(settingsStep: SettingsStep): ModuleWizardStep {
        return object : SdkSettingsStep(settingsStep, this, Condition { sdkTypeId -> isSuitableSdkType(sdkTypeId) }) {
            override fun updateDataModel() {
                super.updateDataModel()
                sourcePaths = emptyList()
            }
        }
    }
}
